ProjFS.Mac/PrjFSKextLogDaemon/PrjFSKextLogDaemon.cpp

```diff
@@ -221,12 +239,141 @@ static bool TryFetchAndLogKextHealthData(io_connect_t connection)
+            "\"FindRootHits\":\"" + to_string(healthData.totalFindRootForVnodeHits) + "\","
+            "\"FindRootMisses\":\"" + to_string(healthData.totalFindRootForVnodeMisses) + "\","
+            "\"RefreshRoot\":\"" + to_string(healthData.totalRefreshRootForVnode) + "\","
+            "\"InvalidateRoot\":\"" + to_string(healthData.totalInvalidateVnodeRoot) + "\","
+        "}");
+        
     }
     else
     {
+        WriteJsonToMessageListener(
+        "{"
+            "\"ErrorMessage\":\"Fetching profiling data from kernel failed\","
+            "\"ret\":\"" + to_string(ret) + "\","
+        "}");
+        
         fprintf(stderr, "fetching profiling data from kernel failed: 0x%x\n", ret);
         return false;
     }
     
     return true;
 }
+
+static void CreatePipeToMessageListener()
+{
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        // Already connected
+        return;
+    }
+
+    s_messageListenerSocket = socket(PF_UNIX, SOCK_STREAM, 0);
+    if (s_messageListenerSocket < 0)
+    {
+        os_log_with_type(
+            s_kextLogger,
+            OS_LOG_TYPE_DEFAULT,
+            "Failed to create a new socket, path: %s, error: %d",
+            s_messageListenerSocketPath.c_str(),
+            errno);
+        
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+        return;
+    }
+    
+    struct sockaddr_un socket_address;
+    memset(&socket_address, 0, sizeof(struct sockaddr_un));
+    
+    socket_address.sun_family = AF_UNIX;
+    size_t resultLength = strlcpy(socket_address.sun_path, s_messageListenerSocketPath.c_str(), sizeof(socket_address.sun_path));
+    
+    if (resultLength >= sizeof(socket_address.sun_path))
+    {
+        os_log_with_type(
+            s_kextLogger,
+            OS_LOG_TYPE_DEFAULT,
+            "Could not copy socket path: %s, insufficient buffer. resultLength: %lu, sizeof(socket_address.sun_path): %lu",
+            s_messageListenerSocketPath.c_str(),
+            resultLength,
+            sizeof(socket_address.sun_path));
+        
+        goto ClosePipeAndCleanup;
+    }
+    
+    if(0 == connect(s_messageListenerSocket, (struct sockaddr *) &socket_address, sizeof(struct sockaddr_un)))
+    {
+        os_log_with_type(
+            s_kextLogger,
+            OS_LOG_TYPE_DEFAULT,
+            "Connected to message listener on socket '%s'",
+            s_messageListenerSocketPath.c_str());
+        return;
+    }
+    
+    os_log_with_type(
+        s_kextLogger,
+        OS_LOG_TYPE_DEFAULT,
+        "Failed to connect socket, pipeName: %s, error: %d",
+        s_messageListenerSocketPath.c_str(),
+        errno);
+    
+ClosePipeAndCleanup:
+
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        close(s_messageListenerSocket);
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+    }
+}
+
+static void WriteJsonToMessageListener(const string& jsonMessage)
+{
+    if (INVALID_SOCKET_FD == s_messageListenerSocket)
```

Comment:
Need to think about thread safety here, `WriteJsonToMessageListener` and `CreatePipeToMessageListener` could be called on separate threads (and there could be multiple calls to `WriteJsonToMessageListener` at the same time on different threads).

Reply:
@wilbaker approach makes sense to me!  Yes probably some locking around connections and message sending.